idaes/apps/grid_integration/multiperiod/clustering.py

```diff
@@ -0,0 +1,240 @@
+
+    Returns:
+        daily_data: pd.DataFrame,
+            Correctly formatted daily LMP data for later use
+    """
+    if horizon_length > len(raw_data):
+        raise ValueError(
+            f"Horizon length {horizon_length} exceeds the price signal length of {len(raw_data)}"
+        )
+
+    elements_ignored = len(raw_data) % horizon_length
+    if elements_ignored:
+        _logger.warning(
+            f"Length of the price signal is not an integer multiple of horizon_length.\n"
+            f"\tIgnoring the last {elements_ignored} elements in the price signal."
+        )
+
+    daily_data = {
+        j: raw_data[((j - 1) * horizon_length) : (j * horizon_length)]
+        for j in range(1, (len(raw_data) // horizon_length) + 1)
+    }
+
+    # DataFrame arranges the data for each day as a column. Since most clustering techniques
+    # require different samples as rows, we take the transpose before returning the data.
+    return pd.DataFrame(daily_data).transpose()
+
+
+def cluster_lmp_data(
+    raw_data: list, horizon_length: int, n_clusters: int, seed: int = 42
+):
+    """
+    Clusters the given price signal into n_clusters using the k-means clustering
+    technique.
+
+    Args:
+        raw_data: list,
+            Columnar data for a given LMP signal.
+
+        horizon_length: int,
+            Length of each cluster (representative day/period)
+
+        n_clusters: int,
+            Number of clusters desired for the data (representative days).
+
+        seed: int,
+            Seed value for initializing random number generator within Kmeans
+
+    Returns:
+        lmp_data_clusters: dict
+            A dictionary of representative day LMP data, indices are indexed
+                by integers starting at 1. Example: ::
+
+                    {1: {1: 4, 2: 3, 3: 5},
+                    2: {1: 1, 2: 7, 3: 3}}
+
+
+        weights: dict
+            A dictionary of weights for each representative day, indexed the
+                same way as lmp_data. Example: ::
+
+                    {1: 45, 2: 56}
+    """
+    # reconfiguring raw data
+    daily_data = generate_daily_data(raw_data, horizon_length)
+
+    # KMeans clustering with the optimal number of clusters
+    kmeans = KMeans(n_clusters=n_clusters, random_state=seed).fit(daily_data)
+    centroids = kmeans.cluster_centers_
+    labels = kmeans.labels_
+
+    # Set any centroid values that are < 1e-4 to 0 to avoid noise
+    centroids = centroids * (abs(centroids) >= 1e-4)
+
+    # Create dicts for lmp data and the weight of each cluster
+    num_days, num_time_periods = centroids.shape
+    lmp_data_clusters = {
+        d + 1: {t + 1: centroids[d, t] for t in range(num_time_periods)}
+        for d in range(num_days)
+    }
+    weights = {d + 1: sum(labels == d) for d in range(num_days)}
+
+    return lmp_data_clusters, weights
+
+
+def get_optimal_num_clusters(
+    samples: Union[pd.DataFrame, np.array],
+    kmin: int = 1,
```

Comment:
Default should be at least 2 since that is the minimum that silhouette requires

Reply:
Good catch! Changed the default to 2.